codec/address/bech32_codec.go

```diff
@@ -61,3 +130,38 @@ func (bc Bech32Codec) BytesToString(bz []byte) (string, error) {
 
 	return text, nil
 }
+
+func (cbc cachedBech32Codec) BytesToString(bz []byte) (string, error) {
+	key := conv.UnsafeBytesToStr(bz)
+	cbc.mu.Lock()
+	defer cbc.mu.Unlock()
+
+	if addr, ok := cbc.cache.Get(key); ok {
+		return addr.(string), nil
+	}
+
+	addr, err := cbc.codec.BytesToString(bz)
+	if err != nil {
+		return "", err
+	}
+	cbc.cache.Add(key, addr)
+
+	return addr, nil
+}
```

Comment:
Can there be more than 1 bech32 prefix per type? The cache is global and may return wrong strings for the byte representation.

Reply:
Most of the time its gonna be just one bech32 prefix. But there may be uses cases with more than one. For example someone who wants to develop a multi chain client. In those cases this cache may be a problem.

I'm more inclined to add the prefix to the key now. Or not share LRUs between codecs at all.

🤔🤔🤔

Reply:
The problem is BytesToString for an address as this path may not point to an unique address.  A unique cache key prefix for each `cachedBech32Codec` instance to the byte key would prevent this. You can keep the global cache but would require a global index for the bech32 prefix to cache prefix as well so that `NewBech32Codec(prefix string)` resolve to the same cache key prefix.
The alternative is a cache per `cachedBech32Codec` as you pointed out. This gives a better isolation but it comes with the cost of extra memory which can be a lot. You would still need a global index but for the bech32 prefix to cache now. 
A third alternative is caching for the default bech32 prefixes defined in Config only. With this approach you won't need indexes or cache key prefixes. I would assume it would cover the majority of use cases on a chain.
Nevertheless, there is complexity to pass these defaults to the cache on initialization.

Reply:
> a global index for the bech32 prefix to cache prefix as well so that `NewBech32Codec(prefix string)` resolve to the same cache key prefix.

I think caches defined prefix should be sufficient, we can concatenate the prefix to the bytes as a key.
https://github.com/cosmos/cosmos-sdk/blob/5a5c63890e5d8b0d7b547697dfa2a0d630a502cc/codec/address/bech32_codec.go#L15-L17

E.g. if two address codecs are defined with prefixes `cosmos` and `osmo` the LRU table'll look like this:
|  Key | Value   | 
|---|---|
| cosmos010101  |  cosmos1g3...  |   
| osmo010101  |  osmo1vt...  |